Unsolicited unused local vals in macro expansions with 2.13.17

First of all, I want to confirm that https://github.com/scala/scala/pull/11080 actually improved the unused check, and now it emits warnings that weren't present in 2.13.16 and earlier. However, this revealed some interesting behaviour in macro expansions. I can't say whether it's a bug, a known bug, or not a bug at all — it'd be better for you to assess that. But in any case, I'd like to glean some wisdom on how to workaround it without resorting to anything radical things like giving up on the `-Xfatal-warnings` and/or `-Ywarn-unused:locals` compiler options.

## Reproduction steps

Scala version: 2.13.17. 

So, imagine we have some macro, we pass an expression to it, then manipulate it in the macro impl, and ultimately include it in the final macro expansion. A minimal repro could be:

```scala
def stub[A](impl: A): Unit = macro Impl.stub0[A]

object Impl {
  def stub0[A](impl: c.Expr[A]): c.Tree = c.untypecheck(impl.tree)  
}
```

On the application side, imagine we're passing an instance of a case class and using named arguments (note that the **reversed order of arguments matters**! the issue disappears if the order is the same as in the case class fields definition, which is also interesting though):

```scala
val randStr = ""

case class Foo(foo: String, flag: Boolean)

stub(Foo(flag = true, foo = randStr))
```

Which gives us a compilation error (it's actually a warning, yes, it's an error just because of my scalac options):

```
local val x$1 in value $anonfun is never used
[error]       stub(Foo(flag = true, foo = randStr))
```

And if we print the `showCode(c.untypecheck(impl.tree))`, we'll determine the troubler maker:

```
{
  val x$1: Boolean(true) = true;
  val x$2: String = randStr;
  Foo.apply(x$2, true)
}
```

## But
That behaviour in macro expansions existed even before 2.13.17 (I checked at least in 2.13.16), but with the adjustment of the unused check in 2.13.17, it started to cause compilation failures (warnings, yes).

I remember the issue that the named arg rewrite leaves a local unused, but I don't recall (and can't find) the ticket.

Here is what I tried (after struggling to remember how to write a Scala macro...):

`t13139-test.scala`
```
//> using scala 2.13.17
//> using dep org.scala-lang:scala-compiler:2.13.17
//> using options -Wunused:_

case class Foo(foo: String, flag: Boolean)

object Main extends App {
  val randStr = ""
  println {
    val stuff = "" // actually unused
    Stub.stub(Foo(flag = true, foo = randStr))
  }
}
```
and `t13139.scala`
```
//> using scala 2.13.17
//> using dep org.scala-lang:scala-compiler:2.13.17

import language.experimental.macros
import scala.reflect.macros.blackbox.Context

object Stub {
  def stub[A](impl: A): Unit = macro stub0[A]

  def stub0[A: c.WeakTypeTag](c: Context)(impl: c.Expr[A]): c.Tree = {
    import c.universe._

    c.untypecheck(impl.tree)
  }
}
```
then
```
scala-cli compile --server=false -Vprint:typer -d /tmp t13139.scala
```
and
```
➜  snips scala-cli compile --server=false -Vprint:typer -d /tmp -cp /tmp t13139-test.scala
/home/amarki/snips/t13139-test.scala:10: warning: local val stuff in value <local Main> is never used
    val stuff = ""
        ^
[[syntax trees at end of                     typer]] // t13139-test.scala
package <empty> {
  case class Foo extends AnyRef with Product with Serializable {
    <caseaccessor> <paramaccessor> private[this] val foo: String = _;
    <stable> <caseaccessor> <accessor> <paramaccessor> def foo: String = Foo.this.foo;
    <caseaccessor> <paramaccessor> private[this] val flag: Boolean = _;
    <stable> <caseaccessor> <accessor> <paramaccessor> def flag: Boolean = Foo.this.flag;
    def <init>(foo: String, flag: Boolean): Foo = {
      Foo.super.<init>();
      ()
    };
    <synthetic> def copy(foo: String = foo, flag: Boolean = flag): Foo = new Foo(foo, flag);
    <synthetic> def copy$default$1: String = Foo.this.foo;
    <synthetic> def copy$default$2: Boolean = Foo.this.flag;
    override <synthetic> def productPrefix: String = "Foo";
    <synthetic> def productArity: Int = 2;
    <synthetic> def productElement(x$1: Int): Any = x$1 match {
      case 0 => Foo.this.foo
      case 1 => Foo.this.flag
      case _ => scala.runtime.Statics.ioobe[Any](x$1)
    };
    override <synthetic> def productIterator: Iterator[Any] = scala.runtime.ScalaRunTime.typedProductIterator[Any](Foo.this);
    <synthetic> def canEqual(x$1: Any): Boolean = x$1.$isInstanceOf[Foo]();
    override <synthetic> def productElementName(x$1: Int): String = x$1 match {
      case 0 => "foo"
      case 1 => "flag"
      case _ => scala.runtime.Statics.ioobe[String](x$1)
    };
    override <synthetic> def hashCode(): Int = {
      <synthetic> var acc: Int = -889275714;
      acc = scala.runtime.Statics.mix(acc, 70822);
      acc = scala.runtime.Statics.mix(acc, scala.runtime.Statics.anyHash(foo));
      acc = scala.runtime.Statics.mix(acc, if (flag)
        1231
      else
        1237);
      scala.runtime.Statics.finalizeHash(acc, 2)
    };
    override <synthetic> def toString(): String = scala.runtime.ScalaRunTime._toString(Foo.this);
    override <synthetic> def equals(x$1: Any): Boolean = Foo.this.eq(x$1.asInstanceOf[Object]).||(x$1 match {
  case (_: Foo) => true
  case _ => false
}.&&({
      <synthetic> val Foo$1: Foo = x$1.asInstanceOf[Foo];
      Foo.this.flag.==(Foo$1.flag).&&(Foo.this.foo.==(Foo$1.foo).&&(Foo$1.canEqual(Foo.this)))
    }))
  };
  <synthetic> object Foo extends scala.runtime.AbstractFunction2[String,Boolean,Foo] with java.io.Serializable {
    def <init>(): Foo.type = {
      Foo.super.<init>();
      ()
    };
    final override <synthetic> def toString(): String = "Foo";
    case <synthetic> def apply(foo: String, flag: Boolean): Foo = new Foo(foo, flag);
    case <synthetic> def unapply(x$0: Foo): Option[(String, Boolean)] = if (x$0.eq(null))
      scala.None
    else
      Some.apply[(String, Boolean)](scala.Tuple2.apply[String, Boolean](x$0.foo, x$0.flag));
    <synthetic> private def writeReplace(): Object = new scala.runtime.ModuleSerializationProxy(classOf[Foo$])
  };
  object Main extends AnyRef with App {
    def <init>(): Main.type = {
      Main.super.<init>();
      ()
    };
    private[this] val randStr: String = "";
    <stable> <accessor> def randStr: String = Main.this.randStr;
    scala.Predef.println({
      val stuff: String = "";
      ({
        <artifact> val x$1: Boolean(true) = true;
        <artifact> val x$2: String = Main.this.randStr;
        {
          Foo.apply(x$2, true);
          ()
        }
      }: Unit)
    })
  }
}

1 warning
```
Please show me a correct reproduction or tell me what I did wrong.

Oh wait a sec, maybe you `-Wmacros:both` instead of `-Wmacros:default` (the default behavior):
```
➜  snips scala-cli compile --server=false -Vprint:typer -d /tmp -cp /tmp t13139-test.scala
/home/amarki/snips/t13139-test.scala:10: warning: local val stuff in value <local Main> is never used
    val stuff = ""
        ^
/home/amarki/snips/t13139-test.scala:11: warning: local val x$1 in value <local Main> is never used
    Stub.stub(Foo(flag = true, foo = randStr))
                         ^
```
The better setting for `-Wmacros` means that the expansion resolves definitions but does not introduce them. It works that way with Scala 3 inlining, which is a similar idea.

The local is an `artifact` but not `synthetic`, which is why it is warnable.

There is a fix for 2.13.18 related to named args and annotations; I'll pursue whether this case is fixed or fixable.
https://github.com/scala/scala/pull/11149/files#r2426794565

--- a/src/compiler/scala/tools/nsc/ast/Trees.scala
+++ b/src/compiler/scala/tools/nsc/ast/Trees.scala
@@ -19,6 +19,8 @@ import java.lang.System.lineSeparator
 import scala.annotation.nowarn
 
 trait Trees extends scala.reflect.internal.Trees { self: Global =>
+  import self.analyzer.OriginalTreeAttachment
+
   // --- additional cases --------------------------------------------------------
   /** Only used during parsing */
   case class Parens(args: List[Tree]) extends Tree {
@@ -323,8 +325,18 @@ trait Trees extends scala.reflect.internal.Trees { self: Global =>
               // The typer does not accept UnApply. Replace it with Apply, which can be retyped.
               case UnApply(Unapplied(Applied(Select(fun, nme.unapply | nme.unapplySeq), _, _)), args) =>
                 Apply(transform(fun), transformTrees(args))
-              case _ =>
-                val dupl = tree.duplicate
+              case tree =>
+                val dupl = {
+                  val original = tree match {
+                    case tree @ Literal(_) =>
+                      tree.attachments.get[OriginalTreeAttachment] match {
+                        case Some(OriginalTreeAttachment(original)) => original
+                        case _ => tree
+                      }
+                    case tree => tree
+                  }
+                  original.duplicate
+                }
                 // Typically the resetAttrs transformer cleans both symbols and types.
                 // However there are exceptions when we cannot erase symbols due to idiosyncrasies of the typer.
                 // vetoXXX local variables declared below describe the conditions under which we cannot erase symbols.

--- a/test/files/pos/t13139/macro_0.scala
+++ b/test/files/pos/t13139/macro_0.scala
@@ -1,0 +1,12 @@
+// using scala 2.13.17
+// using dep org.scala-lang:scala-compiler:2.13.17
+
+import language.experimental.macros
+import scala.reflect.macros.blackbox.Context
+
+object Stub {
+  def stub[A](impl: A): Unit = macro stub0[A]
+
+  def stub0[A: c.WeakTypeTag](c: Context)(impl: c.Expr[A]): c.Tree =
+    c.untypecheck(impl.tree)
+}

--- a/test/files/pos/t13139/test_1.scala
+++ b/test/files/pos/t13139/test_1.scala
@@ -1,0 +1,10 @@
+//> using options -Werror -Wmacros:both -Wunused:_
+
+case class Foo(foo: String, flag: Boolean)
+
+object Main extends App {
+  val randStr = ""
+  println {
+    Stub.stub(Foo(flag = true, foo = randStr))
+  }
+}

